File: src/leetcode/m209/MinimumSizeSubarraySum.go
package m209

import "math"

// LeetCode 209
func minSubArrayLen(target int, nums []int) int {
	slow, fast := 0, 0
	sum := 0
	minLen := math.MaxInt
	for fast < len(nums) {
		sum += nums[fast]
		fast++

		for sum >= target {
			minLen = minInt(minLen, fast-slow)
			sum -= nums[slow]
			slow++
		}
	}

	if minLen == math.MaxInt {
		return 0
	}
	return minLen
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: src/leetcode/m209/MinimumSizeSubarraySum.java
/*
 * Copyright (c) 2024-2025 THGEEK.COM. All rights reserved.
 */

package leetcode.m209;

/**
 * LeetCode 209
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/08/03 17:20
 */
public class MinimumSizeSubarraySum {
    public int minSubArrayLen(int target, int[] nums) {
        int slow = 0;
        int fast = 0;
        int sum = 0;
        int minLen = Integer.MAX_VALUE;
        while (fast < nums.length) {
            sum += nums[fast++];

            while (sum >= target) {
                minLen = Math.min(minLen, fast - slow);
                sum -= nums[slow++];
            }
        }
        return minLen == Integer.MAX_VALUE ? 0 : minLen;
    }
}
